Add "most common base" rewriter

chrhansk opened the pull request:
This is an attempted fix for #298: It replaces a union of types by the most common (i.e., most specific) base of all types in the union.

carljm commented on monkeytype/typing.py:
```diff
@@ -501,6 +502,64 @@ def rewrite_Generator(self, typ):
+    up to the `object` class (no double inheritance)
+    """
+
+    def compute_bases(self, klass):
+        """
+        Compute list of bases of a given class,
+        starting from (but exluding) the root `object`,
+        ending  (and including) the actual class.
+        """
+        bases = []
+
+        curr_klass = klass
+
+        while curr_klass is not object:
+            bases.append(curr_klass)
+
+            curr_bases = curr_klass.__bases__
+
+            if len(curr_bases) != 1:
+                break
+
+            curr_base = curr_bases[0]
+            curr_klass = curr_base
+        return bases[::-1]
```
I think what you want here is just `klass.mro()[1::-1]`? That would also handle multiple-inheritance, which you don't handle here.

carljm replied:
Maybe also add some tests covering multiple inheritance cases.

carljm commented on monkeytype/typing.py:
```diff
@@ -501,6 +502,64 @@ def rewrite_Generator(self, typ):
+
+    def merge_common_bases(self, first_bases, second_bases):
+        merged_bases = []
+
+        min_len = min(len(first_bases), len(second_bases))
+
+        first_bases = first_bases[:min_len]
+        second_bases = second_bases[:min_len]
```
These lines seem unnecessary, as they just duplicate what is already the default behavior of `zip` (if the given iterables are not the same length, "extra" elements in the longer one are ignored)

carljm reviewed (changes requested):
Thanks for the PR! The new rewriter in `typing.py` looks reasonable, modulo a couple comments, and the test for it in `test_typing.py` looks good. All the rest of the changes in this PR seems unrelated to the union-rewriter feature, and should be removed from the PR.

The PR should also add documentation for the new rewriter.

chrhansk commented:
Thanks for the feedback. I added a test case covering multiple inheritance and fixed up the call to the `zip` function. I also renamed the rewrite to make the intentions more clear. 

Regarding the similarity to `mro()`: The issue in fact revolves around multiple inheritance: The mro states that in case a class is derived from two or more bases, the resolution uses the first of these bases. The rewriter does however enable a partial rewrite even in case of multiple inheritance (see the `test_multiple_bases_nontrivial` test case), which would not be possible based on using mro alone

carljm commented on monkeytype/typing.py:
```diff
@@ -501,6 +502,71 @@ def rewrite_Generator(self, typ):
+
+        for klass in klasses:
+            base = self._compute_bases(klass)
+            if base is None:
+                return union
```
It doesn't seem like there is any scenario in which `_compute_bases` can return `None`? So why include this check?

chrhansk replied:
You are right, I removed that check.

carljm reviewed (changes requested):
Ah, yeah, makes sense that we must not reduce to a common base via multiple inheritance.

A few more comments on the code, and also please run `pipenv run black monkeytype` to fix up the formatting using the autoformatter.

Thank you!

chrhansk commented on monkeytype/typing.py:
```diff
@@ -501,6 +502,71 @@ def rewrite_Generator(self, typ):
+                break
+
+            curr_base = curr_bases[0]
+            curr_klass = curr_base
```
I compacted this as well.

chrhansk commented on monkeytype/typing.py:
```diff
@@ -501,6 +502,71 @@ def rewrite_Generator(self, typ):
+            base = self._compute_bases(klass)
+            if base is None:
+                return union
+            bases.append(base)
```
Yep, that was a leftover...

carljm commented:
Code looks good, thanks for the contribution! Do you feel up to adding some documentation as well? It doesn't need to be extensive, just a mention of the new rewriter and a description of what it does. If not, that's fine, I can add the docs myself, but it may be a couple weeks before I have a chance to do that, and I probably won't merge the PR until docs are ready.

chrhansk commented:
I added some docstrings to the rewriter itself as well as an entry in the changelog. Is that sufficient?

carljm commented on CHANGES.rst:
```diff
@@ -4,6 +4,9 @@ Changelog
 main
 ----
 
+* Add rewriter to replace union of derived classes with common
+  base class (not enabled by default)
```
You can also thank yourself here, and mention the issue and PR number :)
```suggestion
* Add ``RewriteMostSpecificCommonBase` type rewriter to replace union of derived classes with common
  base class (not enabled by default.) Merge of #311, fixes #298. Thanks Christoph Hansknecht.
```

carljm reviewed (commented):
Sorry, I wasn't clear enough! The changelog entry and docstrings look great, but what I actually meant was the public-facing documentation (published at https://monkeytype.readthedocs.io/en/stable/) which is in the `docs/` subdirectory. Specifically an entry for the new rewriter should be added to the "Type Rewriters" section at the bottom of https://raw.githubusercontent.com/Instagram/MonkeyType/main/doc/generation.rst

chrhansk commented:
OK, done.

carljm commented on CHANGES.rst:
```diff
@@ -4,6 +4,9 @@ Changelog
 main
 ----
 
+* Add ``RewriteMostSpecificCommonBase` type rewriter to replace union of derived classes with common
```
```suggestion
* Add ``RewriteMostSpecificCommonBase`` type rewriter to replace union of derived classes with common
```

carljm commented on doc/generation.rst:
```diff
@@ -171,6 +171,14 @@ are performed through configured type rewriters.
+  Rewrites e.g. ``Union[Derived1, Derived2]`` to ``Base``, where
+  ``Derived1`` and ``Derived2`` are both subclasses of a common ``Base``.
+  Unions of arbitrarily many subclasses of a common base class are rewritten
+  as a single class, provided that multiple inherit does not occur in the
```
```suggestion
  as a single class, provided that multiple inheritance does not occur in the
```

carljm reviewed (approved):
Thanks, looks great!
